File: gcpu/emulator.py
from __future__ import annotations

import sys
from io import BytesIO
from typing import List, Dict

from gcpu import utils
from gcpu.instructions import InstructionSet

MEMORY_SIZE = 2 ** 7


class EmulatorRuntimeError(Exception): pass


class ExecutionCyclesExceededError(EmulatorRuntimeError): pass


class InvalidInstructionError(EmulatorRuntimeError): pass


class Emulator:

    def __init__(self, instruction_set: InstructionSet):

        self._output_stream = BytesIO()
        self._instruction_set = instruction_set

        self._memory: List[int] = [0] * MEMORY_SIZE
        self.a = 0
        self.pc = 0
        self.fp = 0
        self.sp = 0

        self.reset()

    def reset(self):
        self.a = 0
        self.pc = 0
        self.fp = 0
        self.sp = 0

    def clear_memory(self):
        for i in range(MEMORY_SIZE):
            self._memory[i] = 0

    @property
    def a_upper(self) -> int:
        return (self.a & 0xff00) >> 8

    @property
    def a_lower(self) -> int:
        return self.a & 0x00ff

    def set_all_memory(self, contents: List[int]):
        if len(contents) > MEMORY_SIZE:
            raise EmulatorRuntimeError('Memory this large not supported yet')
        if not all(byte in range(0, 0x100) for byte in contents):
            raise EmulatorRuntimeError('Contents contains value not in 0-0xff')

        self.clear_memory()
        for index, val in enumerate(contents):
            self._memory[index] = val

    def set_memory_at(self, position: int, value: int):
        if position >= MEMORY_SIZE:
            raise EmulatorRuntimeError(f'Trying to access memory at {hex(position)} which is outside memory range')
        self._memory[position] = value

    def get_memory_at(self, position: int) -> int:
        if position >= MEMORY_SIZE:
            raise EmulatorRuntimeError(f'Trying to access memory at {hex(position)} which is outside memory range')
        return self._memory[position]

    def get_and_inc_pc(self) -> int:
        val = self.get_memory_at(self.pc)
        self.pc += 1
        return val

    def print(self, byte: int):
        self._output_stream.write(byte.to_bytes(1, sys.byteorder))
        self._output_stream.flush()

    def push(self, byte: int):
        self.set_memory_at(self.sp, byte)
        self.sp += 1

    def pop(self) -> int:
        self.sp -= 1
        return self.get_memory_at(self.sp)

    def get_output(self, clear: bool = True) -> bytes:
        result = self._output_stream.getvalue()
        if clear:
            self._output_stream.truncate(0)
            self._output_stream.seek(0)
        return result

    def step_single_instruction(self) -> bool:
        """
        Runs a single instruction, return True if the instruction indicated the program should terminate, False otherwise
        :return:
        """
        ins = self._instruction_set[self.get_and_inc_pc()]
        return ins.emulate(self)  # type: ignore

    def run(self, max_clock_cycles: int = 1_000):
        for clock_counter in range(max_clock_cycles):
            should_terminate = self.step_single_instruction()
            if should_terminate:
                break

        else:
            raise ExecutionCyclesExceededError('Maximum execution cycles exceeded, stuck in infinite loop perhaps?')

File: src/main/kotlin/emulator/Emulator.kt
package se.wingez.emulator

import se.wingez.instructions.InstructionSet

open class EmulatorRuntimeError(message: String) : Exception(message)
class EmulatorCyclesExceeded(message: String) : EmulatorRuntimeError(message)
class EmulatorInstructionError(message: String) : EmulatorRuntimeError(message)
class EmulatorInvalidInstructionError(message: String) : EmulatorRuntimeError(message)

open class Emulator(
    val instructionSet: InstructionSet,
    val memorySize: Int = MEMORY_SIZE,
) {
    companion object {
        const val MEMORY_SIZE = 256
    }

    val outputStream = mutableListOf<UByte>()
    val memory = Array<UByte>(memorySize) { 0u }

    var a: UByte = 0u
    var pc: UByte = 0u
    var fp: UByte = 0u
    var sp: UByte = 0u
    var zeroFlag: Boolean = false

    fun reset() {
        a = 0u
        pc = 0u
        fp = 0u
        sp = 0u
        zeroFlag = false
    }

    fun clearMemory() {
        for (i in 0 until memorySize) {
            memory[i] = 0u
        }
    }

    fun setAllMemory(content: Collection<UByte>) {
        if (content.size > memorySize)
            throw EmulatorRuntimeError("Size of program greater than memory size")

        clearMemory()
        content.forEachIndexed { index, byte -> memory[index] = byte }
    }

    fun setMemoryAt(position: Int, value: UByte) {
        if (position !in 0 until memorySize)
            throw EmulatorRuntimeError("Trying to access memory at $position, which is outside memory range")
        memory[position] = value
    }

    fun setMemoryAt(position: UInt, value: UByte) {
        setMemoryAt(position.toInt(), value)
    }

    fun getMemoryAt(position: Int): UByte {
        if (position !in 0 until memorySize)
            throw EmulatorRuntimeError("Trying to access memory at $position, which is outside memory range")
        return memory[position]
    }

    fun getMemoryAt(position: UInt): UByte {
        return getMemoryAt(position.toInt())
    }

    fun getMemoryAt(position: UByte): UByte {
        return getMemoryAt(position.toInt())
    }

    fun getIncPC(): UByte {
        return getMemoryAt((pc++).toInt())
    }

    fun print(value: UByte) {
        outputStream.add(value)
    }

    fun push(value: UByte) {
        sp--
        setMemoryAt(sp.toInt(), value)
    }

    fun pop(): UByte {
        return getMemoryAt((sp++).toInt())
    }

    fun stepSingleInstruction(): Boolean {
        /**
        Runs a single instruction, return True if the instruction indicated the program should terminate, False otherwise
        :return:
         */
        val ins = instructionSet.instructionFromID(getIncPC())
        return ins.emulate.invoke(this)
    }

    fun run(maxClockCycles: Int = 1000) {
        for (clockCounter in 0 until maxClockCycles) {
            val shouldTerminate = stepSingleInstruction()
            if (shouldTerminate) {
                return
            }
        }
        throw EmulatorCyclesExceeded("Maximum execution cycles exceeded, stuck in infinite loop perhaps?")
    }
}